src/main/drivers/compass/compass_ist8310.c

```diff
@@ -0,0 +1,199 @@
+ *         0   |   0  |  0   |  1   |   Single measurement mode
+ *                                       Others: Reserved
+ *
+ * CTRL_REGB: Control Register 2
+ * Read Write
+ * Default value: 0x0B
+ * 7:4  0   Reserved.
+ * 3    DREN : Data ready enable control:
+ *      0: disable
+ *      1: enable
+ * 2    DRP: DRDY pin polarity control
+ *      0: active low
+ *      1: active high
+ * 1    0   Reserved.
+ * 0    SRST: Soft reset, perform Power On Reset (POR) routine
+ *      0: no action
+ *      1: start immediately POR routine
+ *      This bit will be set to zero after POR routine
+ */
+
+#define IST8310_REG_DATA 0x03
+#define IST8310_REG_WHOAMI 0x00
+
+// I2C Contorl Register
+#define IST8310_REG_CNTRL1 0x0A
+#define IST8310_REG_CNTRL2 0x0B
+#define IST8310_REG_AVERAGE 0x41
+#define IST8310_REG_PDCNTL 0x42
+
+// Parameter
+// ODR = Output Data Rate, we use single measure mode for getting more data.
+#define IST8310_ODR_SINGLE 0x01
+#define IST8310_ODR_10_HZ 0x03
+#define IST8310_ODR_20_HZ 0x05
+#define IST8310_ODR_50_HZ 0x07
+#define IST8310_ODR_100_HZ 0x06
+
+// Device ID (ist8310 -> 0x10)
+#define IST8310_CHIP_ID 0x10
+#define IST8310_AVG_16  0x24
+#define IST8310_PULSE_DURATION_NORMAL 0xC0
+
+#define IST8310_CNTRL2_RESET 0x01
+#define IST8310_CNTRL2_DRPOL 0x04
+#define IST8310_CNTRL2_DRENA 0x08
+
+static bool ist8310Init(magDev_t *magDev)
+{
+    extDevice_t *dev = &magDev->dev;
+
+    busWriteRegister(dev, IST8310_REG_CNTRL1, IST8310_ODR_50_HZ);
+    delay(5);
+
+    busWriteRegister(dev, IST8310_REG_AVERAGE, IST8310_AVG_16);
+    delay(5);
+
+    busWriteRegister(dev, IST8310_REG_PDCNTL, IST8310_PULSE_DURATION_NORMAL);
+    delay(5);
+
+    return true;
+}
+
+static bool ist8310Read(magDev_t * magDev, int16_t *magData)
+{
+    uint8_t buf[6];
+    uint8_t LSB2FSV = 3; // 3mG - 14 bit
+
+    extDevice_t *dev = &magDev->dev;
+
+    // set magData to zero for case of failed read
+    magData[X] = 0;
+    magData[Y] = 0;
+    magData[Z] = 0;
+
+    if (!busReadRegisterBuffer(dev, IST8310_REG_DATA, buf, 6)) {
+        return false;
+    }
+
+    // Looks like datasheet is incorrect and we need to invert Y axis to conform to right hand rule
+    magData[X] =  (int16_t)(buf[1] << 8 | buf[0]) * LSB2FSV;
+    magData[Y] = -(int16_t)(buf[3] << 8 | buf[2]) * LSB2FSV;
+    magData[Z] =  (int16_t)(buf[5] << 8 | buf[4]) * LSB2FSV;
+
+    return true;
+}
+
+#define DETECTION_MAX_RETRY_COUNT 10
+static bool deviceDetect(magDev_t * magDev)
+{
+    extDevice_t *dev = &magDev->dev;
+
+    for (int retryCount = 0; retryCount < DETECTION_MAX_RETRY_COUNT; retryCount++) {
+        delay(10);
+
+        uint8_t sig = 0;
+        bool ack = busReadRegisterBuffer(dev, IST8310_REG_WHOAMI, &sig, 1);
+
+        if (ack && sig == IST8310_CHIP_ID) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+bool ist8310Detect(magDev_t * magDev)
+{
+    extDevice_t *dev = &magDev->dev;
+
+    busDeviceRegister(dev);
```

Comment:
I think that is done at the "init" method in other drivers files. I suppose it has not too much sense to register it if the detection fails.

Reply:
Moved to init :+1: